silq/analysis/analysis.py

```diff
@@ -715,6 +717,38 @@ def analyse_EPR(empty_traces: np.ndarray,
             'mean_high_blip_duration': results_read['mean_high_blip_duration']}
 
 
+def analyse_threshold_up_proportion(up_proportions_arrs: np.ndarray,
+                                    shots_per_frequency: int):
+    proportion_space = np.linspace(0, 1, num=shots_per_frequency + 1,
+                                   endpoint=True)
+    up_proportions_arrs = up_proportions_arrs.reshape(1, -1)
+    kernel = gaussian_kde(up_proportions_arrs)
+    gaussian_up_proportions = kernel(proportion_space)
+    peak_idxs = peakutils.peak.indexes(gaussian_up_proportions,
+                                       thres=0.5/up_proportions_arrs.shape[-1],
+                                       min_dist=shots_per_frequency/5)
```

Comment:
Generally, I agree, but I saw the situations when the spin-up proportion is low (spread from 0.1 to 0.4) and there was a "dip" (missing shots) somewhere around 0.25. In this case according to the code it will define threshold as 0.25, which actually should be 0.5 or so... So I would be careful with this...

Reply:
In that case, it sounds like the smoothing routine didn't work correctly. It should be able to smooth out missing datapoints adequately.

Just putting some of my stats thoughts down here for my future reference:
The standard deviation of a binomial distribution under the CLT is `sigma/sqrt(n)` where `n = samples` and `sigma` is the intrinsic standard deviation of the binomial distribution: `p(1-p)` where `p` is the centre of the peak, e.g. 0.2 or 0.8. The largest value for `sigma` is when `p = 0.5`, so `sigma = 0.25`. I think a conservative choice for the gaussian_kde bandwidth should be `0.25/sqrt(n)`, taking care to note that n is NOT the `shots_per_frequency` but the number of samples gathered for the up-proportions of this round.
We could also think about introducing a confidence level on the threshold found, which would have to be determined somehow. This could allow the use of past threshold with their confidences as weights to be used to inform the current threshold. This is particularly useful if you're using very few samples.

Reply:
if there is a dip - there is dip, and it will be present in density distribution, I am not sure this can be smoothed much - just the depth will change, but the dip will be there...